### DISCUSSION TITLE
Slow Performance of Linq Where statement

### DISCUSSION THREAD

**Discussion by MohanramAnbukkarasu** [dotnet/winforms#4713] [Q&A]
Hi,

I have an ObservableCollection with 4 lakhs unique items. On a button click event,  I execute a Linq Where query to set value for a bool value in the items of the collection. 
`
List<DataClass> DistinctCollection = new List<DataClass>();
List<DataClass> filteredCollection = new List<DataClass>();

for(int i=1;i<=400000; i++)
 {
     DistinctCollection.Add(new DataClass() { Content = i.ToString() });
 }

 filteredCollection = DistinctCollection.Where(x => x.Content.Contains("2")).ToList();

private void button1_Click(object sender, EventArgs e)
{
    // The problem occurs here.
    DistinctCollection.Where(x => filteredCollection.Contains(x)).Select(x => x.IsSelected = true).ToList();
    DistinctCollection.Where(x => !filteredCollection.Contains(x)).Select(x => x.IsSelected = false).ToList();
}

`

Please use below given sample application to reproduce the problem.

The query works properly when the number of items in the collection is less. But the UI freeze when executing the query if the collection contains 4 lakhs and above items. Is there any possible way to improve the performance of this query?

Thanks in advance.

[Linq.Where_Demo.zip](https://github.com/dotnet/winforms/files/6183604/Linq.Where_Demo.zip)

**Comment by RussKie** [comment]
You're executing your operations on the UI thread, thus blocking it from processing it other important messages such as repaint. This essentially makes your app look unresponsive.

Long running and/or CPU heavy work should be performed on a background thread, and marshaled back to the UI thread when necessary, e.g. to update the UI with the progress. 
There are multitude of ways of doing this, e.g. with `async/await`, [`Invoke`](https://docs.microsoft.com/dotnet/api/system.windows.forms.control.invoke) or [`BeginInvoke`](https://docs.microsoft.com/dotnet/api/system.windows.forms.control.begininvoke), [`BackgroundWorker`](https://docs.microsoft.com/dotnet/api/system.componentmodel.backgroundworker), etc.

You may find these resources useful:
* https://docs.microsoft.com/dotnet/csharp/programming-guide/concepts/async/
* https://docs.microsoft.com/dotnet/visual-basic/programming-guide/concepts/async/
* https://stephenhaunts.com/2014/10/14/using-async-and-await-to-update-the-ui-thread/

**Reply by MohanramAnbukkarasu** [reply]
HI @RussKie ,

Sorry for the late reply. 

I have tried as in the below code snippet.


private void button1_Click(object sender, EventArgs e)
        {
            this.BeginInvoke(new Action(() =>
            {
                // The problem occurs here.
                DistinctCollection.Where(x => filteredCollection.Contains(x)).Select(x => x.IsSelected = true).ToList();
                DistinctCollection.Where(x => !filteredCollection.Contains(x)).Select(x => x.IsSelected = false).ToList();
            }));
        }


But still the UI freeze. Please let me know if there is any other way to resolve this. 

Regards,
Mohanram A
